Store short strings inside string's body

MasterDuke17 opened the pull request:
Uses the 8 bytes that the pointer to a buffer usually takes and puts up to eight 8-bit or two 32-bit characters into it.

Started by @timo++.

The time reported by micro-benchmark `MVM_SPESH_BLOCKING=1 nqp-m -e 'my str $s; my int $i := 0; my $n := nqp::time; while $i++ < 10_000_000 { $s := $i }; say(nqp::div_n(nqp::time - $n, 1000000000e0)); say($s)'` drops from ~0.270s to ~0.145s, the number of instructions (for 1m iterations) drops from ~628m to ~555m, and the number of allocation functions called as reported by heaptrack drops from ~10m to ~252k.

However, another micro-benchmark `MVM_SPESH_BLOCKING=1 nqp-m -e 'my str $s := "ab.cd.ef.gh.ij.kl.mn.op.qr.st.uv.wx.yz"; my int $i := 0; my @a; nqp::setelems(@a, 30); my $n := nqp::time; while $i++ < 1_000_000 { @a := nqp::split(".", $s); }; say(nqp::div_n(nqp::time - $n, 1000000000e0)); say(@a[0])'` has no significant change in time reported, and the instruction count increased from ~10.2b to ~11.9b, even though the allocation functions called as reported by heaptrack drops from ~15.2m to ~2.2m.

MasterDuke17 commented:
Not every single case where an MVMString is created do I check whether it could be in-situ. Some of them seemed a bit unlikely to hit in enough instances to be worth the extra checking. However, some might need metrics gathered on the resulting sizes to really know for sure whether it's worth it. And I'd appreciate any suggestions for places I've missed.

MasterDuke17 commented:
The CI fails are all because of t/02-rakudo/reproducible-builds.t, I'm not quite sure what the problem is or how to fix it.

MasterDuke17 commented:
> The CI fails are all because of t/02-rakudo/reproducible-builds.t, I'm not quite sure what the problem is or how to fix it.

Well, locally a re-bootstrap (and then even another just for good luck) didn't fix it.

lizmat commented:
FWIW, I've seen `t/02-rakudo/reproducible-builds.t` flap, but rarely.  Possibly this PR is exposing the cause of this flapper.

timo commented:
on the other hand, if we have such a very small substring, it's also possible to decide "we should store the string in-situ instead of a rope-based substring" ... ok i very vaguely recall that we might already be doing something similar to that?? maybe only for comb for example

timo commented on src/strings/ops.c:
```diff
@@ -337,6 +389,15 @@ static void iterate_gi_into_string(MVMThreadContext *tc, MVMGraphemeIter *gi, MV
+            result->body.storage.in_situ_32[i] = old[i];
+        }
+        result->body.storage_type    = MVM_STRING_IN_SITU_32;
+        MVM_free(old);
```
it would be interesting to see how often this triggers; should we be able to know `result->body.num_graphs` near the start of the function, and could we then always never go with storage type string_grapheme_32 throughout the body of the function?

MasterDuke17 replied:
You're right, we could probably do it right around https://github.com/MoarVM/MoarVM/pull/1802/files#diff-bc76be5614df2473912916db269fc246c79efbc83a0a065ee8b785aa3c68d040R330-R333

MasterDuke17 commented:
> for the second benchmark you posted, it's probably important to note that `@a := nqp::split(...)` will create a new array each go-around, which is probably a big chunk of allocations, and i'd wager that the split will create ropes for all the entries rather than use any real storage. it would be kind of fiddly, but possible, to have in-situ-storage for "non-repeated substring of just a single other string" ropes? with a specific storage type, the pointer can point at the other string, the spot for the strand count can be re-used for a start offset, the num_graphs would naturally double as the "how long is the substring" count. then the split example should be able to reduce its allocations.

> on the other hand, if we have such a very small substring, it's also possible to decide "we should store the string in-situ instead of a rope-based substring" ... ok i very vaguely recall that we might already be doing something similar to that?? maybe only for comb for example

The second thing is (should be) happening now, though kind of in a round-about way that might not be more efficient. Here https://github.com/MoarVM/MoarVM/pull/1802/files#diff-bc76be5614df2473912916db269fc246c79efbc83a0a065ee8b785aa3c68d040R827 in `MVM_string_substring` I added the `&& result->body.num_graphs > 8`, so if it's <= 8 then it goes into the branch that just does `iterate_gi_into_string()`, which should put stuff directly into an in-situ string if it can fit.

> another thing from your benchmark code that i can't determine whether it's useful or not is to have a variant of nqp::split that takes an existing array and puts its data into it, instead of creating a new array to return. this would want a survey of the whole nqp and rakudo code bases to find spots where we can benefit. i don't expect it's very many, though

ISTR some conversation with you and/or lizmat about this, maybe for storing profile information, or rusage? Ah, looks like it was just done directly in the nqp::getrusage op. https://github.com/Raku/nqp/blob/main/docs/ops.markdown#getrusage

timo commented:
```
From e215e74bba2853cde0c9ddd53b1a1a424d860674 Mon Sep 17 00:00:00 2001
From: Timo Paulssen <timonator@perpetuum-immobile.de>
Date: Fri, 5 Jul 2024 22:31:16 +0200
Subject: [PATCH] fix comparison of strings involving in situ strings

---
 src/strings/ops.c | 47 ++++++++++++++++++++++++++---------------------
 1 file changed, 26 insertions(+), 21 deletions(-)

diff --git a/src/strings/ops.c b/src/strings/ops.c
index a7b00465b..32e864152 100644
--- a/src/strings/ops.c
+++ b/src/strings/ops.c
@@ -2454,6 +2454,8 @@ MVMint64 MVM_string_compare(MVMThreadContext *tc, MVMString *a, MVMString *b) {
     MVMStringIndex alen, blen, i = 0, scanlen;
     MVMGraphemeIter gi_a, gi_b;
 
+    MVMuint8 a_is_eight, b_is_eight, a_is_in_situ, b_is_in_situ;
+
     MVM_string_check_arg(tc, a, "compare");
     MVM_string_check_arg(tc, b, "compare");
 
@@ -2468,35 +2470,26 @@ MVMint64 MVM_string_compare(MVMThreadContext *tc, MVMString *a, MVMString *b) {
     /* Otherwise, need to scan them. */
     scanlen = blen < alen ? blen : alen;
 
+    a_is_eight = a->body.storage_type == MVM_STRING_GRAPHEME_8 || a->body.storage_type == MVM_STRING_GRAPHEME_ASCII || a->body.storage_type == MVM_STRING_IN_SITU_8;
+    b_is_eight = b->body.storage_type == MVM_STRING_GRAPHEME_8 || b->body.storage_type == MVM_STRING_GRAPHEME_ASCII || b->body.storage_type == MVM_STRING_IN_SITU_8;
+
+    a_is_in_situ = a->body.storage_type == MVM_STRING_IN_SITU_8 || a->body.storage_type == MVM_STRING_IN_SITU_32;
+    b_is_in_situ = b->body.storage_type == MVM_STRING_IN_SITU_8 || b->body.storage_type == MVM_STRING_IN_SITU_32;
+
     /* Short circuit a case where the other conditionals won't speed it up */
     if (a->body.storage_type == MVM_STRING_STRAND || b->body.storage_type == MVM_STRING_STRAND) {
         /* do nothing */
     }
-    else if ((a->body.storage_type == MVM_STRING_GRAPHEME_8 || a->body.storage_type == MVM_STRING_GRAPHEME_ASCII)
-          && (b->body.storage_type == MVM_STRING_GRAPHEME_8 || b->body.storage_type == MVM_STRING_GRAPHEME_ASCII)) {
-        MVMGrapheme8  *a_blob8 = a->body.storage.blob_8;
-        MVMGrapheme8  *b_blob8 = b->body.storage.blob_8;
+    else if (a_is_eight && b_is_eight) {
+        MVMGrapheme8  *a_blob8 = a_is_in_situ ? a->body.storage.in_situ_8 : a->body.storage.blob_8;
+        MVMGrapheme8  *b_blob8 = b_is_in_situ ? b->body.storage.in_situ_8 : b->body.storage.blob_8;
         while (i < scanlen && a_blob8[i] == b_blob8[i]) {
             i++;
         }
     }
-    else if (a->body.storage_type == MVM_STRING_IN_SITU_8 || b->body.storage_type == MVM_STRING_IN_SITU_8) {
-        MVMGrapheme8  *a_blob8 = a->body.storage.in_situ_8;
-        MVMGrapheme8  *b_blob8 = b->body.storage.in_situ_8;
-        while (i < scanlen && a_blob8[i] == b_blob8[i]) {
-            i++;
-        }
-    }
-    else if (a->body.storage_type == MVM_STRING_GRAPHEME_32 && b->body.storage_type == MVM_STRING_GRAPHEME_32) {
-        MVMGrapheme32  *a_blob32 = a->body.storage.blob_32;
-        MVMGrapheme32  *b_blob32 = b->body.storage.blob_32;
-        while (i < scanlen && a_blob32[i] == b_blob32[i]) {
-            i++;
-        }
-    }
-    else if (a->body.storage_type == MVM_STRING_IN_SITU_32 || b->body.storage_type == MVM_STRING_IN_SITU_32) {
-        MVMGrapheme32  *a_blob32 = a->body.storage.in_situ_32;
-        MVMGrapheme32  *b_blob32 = b->body.storage.in_situ_32;
+    else if (!a_is_eight && !b_is_eight) {
+        MVMGrapheme32  *a_blob32 = a_is_in_situ ? a->body.storage.in_situ_32 : a->body.storage.blob_32;
+        MVMGrapheme32  *b_blob32 = b_is_in_situ ? b->body.storage.in_situ_32 : b->body.storage.blob_32;
         while (i < scanlen && a_blob32[i] == b_blob32[i]) {
             i++;
         }
@@ -2509,9 +2502,15 @@ MVMint64 MVM_string_compare(MVMThreadContext *tc, MVMString *a, MVMString *b) {
             case MVM_STRING_GRAPHEME_ASCII:
                 blob8 = a->body.storage.blob_8;
                 break;
+            case MVM_STRING_IN_SITU_8:
+                blob8 = a->body.storage.in_situ_8;
+                break;
             case MVM_STRING_GRAPHEME_32:
                 blob32 = a->body.storage.blob_32;
                 break;
+            case MVM_STRING_IN_SITU_32:
+                blob32 = a->body.storage.in_situ_32;
+                break;
             default:
                 MVM_exception_throw_adhoc(tc,
                     "String corruption in string compare. Unknown string type.");
@@ -2521,9 +2520,15 @@ MVMint64 MVM_string_compare(MVMThreadContext *tc, MVMString *a, MVMString *b) {
             case MVM_STRING_GRAPHEME_ASCII:
                 blob8 = b->body.storage.blob_8;
                 break;
+            case MVM_STRING_IN_SITU_8:
+                blob8 = b->body.storage.in_situ_8;
+                break;
             case MVM_STRING_GRAPHEME_32:
                 blob32 = b->body.storage.blob_32;
                 break;
+            case MVM_STRING_IN_SITU_32:
+                blob32 = b->body.storage.in_situ_32;
+                break;
             default:
                 MVM_exception_throw_adhoc(tc,
                     "String corruption in string compare. Unknown string type.");
-- 
2.45.1
```

this fixes reproducible builds

timo reviewed (approved):
i think it's ready to merge
